synapse/appservice/__init__.py

```diff
@@ -172,15 +172,11 @@ async def _matches_user_in_member_list(
         Returns:
             True if this service would like to know about this room.
         """
-        member_list = await store.get_users_in_room(
-            room_id, on_invalidate=cache_context.invalidate
+        app_service_users_in_room = await store.get_app_service_users_in_room(
+            room_id, self, on_invalidate=cache_context.invalidate
         )
 
-        # check joined member events
-        for user_id in member_list:
-            if self.is_interested_in_user(user_id):
-                return True
-        return False
+        return len(app_service_users_in_room) > 0
```

Comment:
As [@clokep mentioned](https://matrix.to/#/!SGNQGPGUwtcPBUotTL:matrix.org/$feRc-xR0MZ5e0sZjMrJIJgwFIJtIbhrK9zL7zHsnIKM?via=jki.re&via=matrix.org&via=element.io), `_matches_user_in_member_list` and `get_app_service_users_in_room` are suspiciously almost the same so we could re-use one in the other.

This technically takes more time the first time we hit this spot because before we would bail early as soon as we found one user that the appservice was interested in. But now we're at least re-using the `get_app_service_users_in_room` function which is also cached.

The only problem is that it looks like `get_app_service_users_in_room` is only used elsewhere in `_compute_msc3202_otk_counts_and_fallback_keys` so maybe it's not that common to use the cache. We're doing more work and barely using it later 🤔

Reply:
Reverted back to having them separate but added a comment about a future possibility of re-using when `get_app_service_users_in_room` is used more so the juice is worth the squeeze.

Reply:
Oh the bailing early is a good call that this likely doesn't make sense. 👍